File: CSharpCodeBase/aiactions/ai_interactenemy.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_interactenemy :ainode{
 // Attack enemy
 public void init (self, entity);
   AINode.init(self, entity);
 }})
 public void Visit(){
   if(this.status == NODE_READY  ){ 
     this.timeout = this.entity.weaponContainer:GetAttackTimeout(this.entity.weaponContainer.interactWeapon);
     this.status = NODE_RUNNING;
     this.attack = false;
     // start attacking
     this.entity.combat:Attack(this.parent.selected_target, this.entity.weaponContainer.interactWeapon);
   }
   
   if(this.status == NODE_RUNNING  &&  this.attack  ){ 
     // calc timeout
     this.timeout = this.timeout - GameController.deltaTime;
     if(this.timeout <= 0  ){ 
       this.status = NODE_SUCCESS;
     }
   }
   
   return this.status;
 }
 public void OnEvent(data){
   if(data == "punch"  ){        
     this.attack = true;
     if(this.parent.selected_target.interactable  &&  this.entity.weaponContainer:CanAttack(this.parent.selected_target, this.entity.weaponContainer.interactWeapon)  ){ 
       var canBeFood = true;
       if(this.parent.selected_target.CanBeFood  ){ 
         canBeFood = this.parent.selected_target:CanBeFood();
       }
       if(canBeFood  ){ 
         this.parent.selected_target:Message("OnInteract", this.entity);
       }
       this.entity.starvation:SetState(false);
     }        
   }
 }
 }}

File: CSharpCodeBase/aiactions/ai_interactenemy.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'

-- Attack enemy
local AI_InteractEnemy = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
end})

function AI_InteractEnemy:Visit()
  if self.status == NODE_READY then
    self.timeout = self.entity.weaponContainer:GetAttackTimeout(self.entity.weaponContainer.interactWeapon)
    self.status = NODE_RUNNING
    self.attack = false
    -- start attacking
    self.entity.combat:Attack(self.parent.selected_target, self.entity.weaponContainer.interactWeapon)
  end
  
  if self.status == NODE_RUNNING and self.attack then
    -- calc timeout
    self.timeout = self.timeout - GameController.deltaTime
    if self.timeout <= 0 then
      self.status = NODE_SUCCESS
    end
  end
  
  return self.status
end

function AI_InteractEnemy:OnEvent(data)
  if data == "punch" then       
    self.attack = true
    if self.parent.selected_target.interactable and self.entity.weaponContainer:CanAttack(self.parent.selected_target, self.entity.weaponContainer.interactWeapon) then
      local canBeFood = true
      if self.parent.selected_target.CanBeFood then
        canBeFood = self.parent.selected_target:CanBeFood()
      end
      if canBeFood then
        self.parent.selected_target:Message("OnInteract", self.entity)
      end
      self.entity.starvation:SetState(false)
    end        
  end
end

return AI_InteractEnemy